File: set3/4/main.py
import sys

def input():
    return sys.stdin.readline().rstrip()

def erase_simple_tag(data: str) -> str:
    ret = ""
    space = True
    bucket_open = False
    for ch in data:
        if ch == '>':
            bucket_open = False
            continue
        if ch == '<':
            bucket_open = True
            continue

        if bucket_open:
            continue

        if ch == ' ':
            if space: continue
            space = True
            ret = f"{ret}{ch}"
        else:
            ret = f"{ret}{ch}"
            space = False
    return ret.rstrip()

html = input()
L, N = 0, len(html)
while L < N:
    idx = html.find("title=\"", L)
    if idx == -1:
        break
    end_title = html.find("\">", idx)
    end_tag   = html.find("</div>", idx)
    title     = html[idx + 7:end_title]
    print(f"title : {title}")

    cursor = L
    while True:
        start_p_tag = html.find("<p>", cursor)
        if start_p_tag == -1 or start_p_tag > end_tag: break
        end_p_tag = html.find("</p>", start_p_tag)
        paragraph = html[start_p_tag + 3:end_p_tag]
        print(erase_simple_tag(paragraph))
        cursor = end_p_tag
    L = end_tag

File: set3/4/main.cpp
#include<iostream>
#include<string>

using namespace std;

string erase_simple_tag(string data) {
    string ret;
    bool space = true;
    bool bucket_open = false;
    for(int i = 0; i < (int)data.size(); i++) {
        if(data[i] == '>') {
            bucket_open = false;
            continue;
        }
        if(data[i] == '<') {
            bucket_open = true;
            continue;
        }
        if(bucket_open) continue;
        if(data[i] == ' ') {
            if(space) continue;
            space = true;
            ret += data[i];
        }
        else {
            ret += data[i];
            space = false;
        }
    }
    while(ret.back() == ' ') ret.pop_back();
    return ret;
}

int main(){
    ios::sync_with_stdio(false); cin.tie(0);

    string html; getline(cin, html);
    int L = 0, N = html.size();
    while(L < N) {
        auto idx = html.find("title=\"", L);
        if(idx == string::npos) break;
        auto end_title = html.find("\">", idx);
        auto end_tag = html.find("</div>", idx);
        string title  = html.substr(idx + 7, end_title - idx - 7);
        cout << "title : " << title << '\n';

        int cursor = L;
        while(true) {
            auto start_p_tag = html.find("<p>", cursor);
            if(start_p_tag == string::npos) break;
            if(start_p_tag  > end_tag) break;
            auto end_p_tag   = html.find("</p>", start_p_tag);
            string paragraph = html.substr(start_p_tag + 3, end_p_tag - start_p_tag - 3);
            cout << erase_simple_tag(paragraph) << '\n';
            cursor = end_p_tag;
        }
        L = end_tag;
    }
    return 0;
}